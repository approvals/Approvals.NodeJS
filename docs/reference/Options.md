<a id="top"></a>

# Options

<!-- toc -->
## Contents

  * [Links](#links)<!-- endToc -->

## Links

- [How to Configure Reporters with Options](../how_tos/ConfigureReportersWithOptions.md)

---

[Back to User Guide](/doc/README.md#top)
