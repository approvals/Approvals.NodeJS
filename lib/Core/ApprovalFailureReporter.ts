export interface ApprovalFailureReporter {
    name: string;

    canReportOn(): boolean;

    report(approvedFilePath: string, receivedFilePath: string): void;
}
