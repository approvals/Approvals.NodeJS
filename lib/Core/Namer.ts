export interface Namer {
  getApprovedFile(ext: string): string;
  getReceivedFile(ext: string): string;
}
