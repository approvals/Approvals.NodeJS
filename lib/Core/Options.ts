class FileOptions {
  private options: Options;
  constructor(options: Options) {
    this.options = options;
  }

  withFileExtention(extensionWithDot: string): Options {
    return this.options.modify("FileExtention", extensionWithDot);

  }

  getFileExtension(): string {
    return this.options.get("FileExtention", () => ".txt")
  }
}

export class Options {
  protected fields: { [n: string]: any };
  constructor() {
    this.fields = {}
  }

  modify(key: string, value: any): Options {
    const next = new Options();
    for (const key1 in next.fields) {
      next.fields[key1] = this.fields[key1]
    }
    next.fields[key] = value;
    return next;
  }
  forFile(): FileOptions {
    return new FileOptions(this);
  }

  get<Type>(key: string, default1: () => Type): Type {
    if (this.fields[key] == undefined) {
      return default1();
    }
    return this.fields[key]

  }
}