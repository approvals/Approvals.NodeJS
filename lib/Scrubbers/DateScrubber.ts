import {Scrubbers} from "./Scrubbers";
import type {Scrubber} from "./Scrubbers";

class DateFormat {
    public regex: RegExp;
    public examples: string[];

    constructor(regex: RegExp, examples: string[]) {
        this.regex = regex;
        this.examples = examples;
    }
}

function __(regExp: RegExp, strings: string[]): DateFormat {
    return new DateFormat(regExp, strings);
}

export class DateScrubber {
    static getSupportedFormats(): DateFormat[] {
        return [
            __(/[a-zA-Z]{3} [a-zA-Z]{3} \d{2} \d{2}:\d{2}:\d{2}/g,
                ["Tue May 13 16:30:00"])
            , __(/[a-zA-Z]{3} [a-zA-Z]{3} \d{2} \d{2}:\d{2}:\d{2} [a-zA-Z]{3,4} \d{4}/g,
                ["Wed Nov 17 22:28:33 EET 2021"])
            , __(/[a-zA-Z]{3} [a-zA-Z]{3} \d{2} \d{4} \d{2}:\d{2}:\d{2}.\d{3}/g,
                ["Tue May 13 2014 23:30:00.789"])
            , __(/[a-zA-Z]{3} [a-zA-Z]{3} \d{2} \d{2}:\d{2}:\d{2} -\d{4} \d{4}/g,
                ["Tue May 13 16:30:00 -0800 2014"])
            , __(/\d{2} [a-zA-Z]{3} \d{4} \d{2}:\d{2}:\d{2},\d{3}/g, ["13 May 2014 23:50:49,999"])
            , __(/[a-zA-Z]{3} \d{2}, \d{4} \d{2}:\d{2}:\d{2} [a-zA-Z]{2} [a-zA-Z]{3}/g, ["May 13, 2014 11:30:00 PM PST"])
            , __(/\d{2}:\d{2}:\d{2}/g, ["23:30:00"])
            , __(/\d{4}\/\d{2}\/\d{2} \d{2}:\d{2}:\d{2}.\d{2}\d/g, ["2014/05/13 16:30:59.786"])
            , __(/\d{4}-\d{1,2}-\d{1,2}T\d{1,2}:\d{2}Z/g, [
                "2020-9-10T08:07Z",
                "2020-09-9T08:07Z",
                "2020-09-10T8:07Z",
                "2020-09-10T08:07Z",
            ])
            , __(/\d{4}-\d{1,2}-\d{1,2}T\d{1,2}:\d{2}:\d{2}Z/g, ["2020-09-10T08:07:89Z"])
            , __(/\d{4}-\d{1,2}-\d{1,2}T\d{1,2}:\d{2}\:\d{2}\.\d{3}Z/g, ["2020-09-10T01:23:45.678Z"])
            , __(/\d{8}T\d{6}Z/g, ["20210505T091112Z"])
            , __(/[a-zA-Z]{3}, \d{2} [a-zA-Z]{3} \d{4} \d{2}:\d{2}:\d{2} [a-zA-Z]{3}/g, ["Sun, 06 Nov 2022 11:23:20 GMT"])
        ];
    }

    static getScrubberFor(example: string): Scrubber {
        let supported = ""
        for (let format of DateScrubber.getSupportedFormats()) {
            const scrubber = this.create(format.regex);
            const scrubbed = scrubber(example);
            if (scrubbed === "<date_1>") {
                return scrubber;
            }
            supported += `    ${format.examples[0]} | ${format.regex} \n`;
        }
        throw new Error(`No match found for '${example}'.\n Feel free to add your date at https://github.com/approvals/Approvals.NodeJS/issues/139 \n Current supported formats are: \n${supported}`)
    }

    public static create(regex: RegExp): Scrubber {
        return Scrubbers.createReqexScrubber(regex, t => `<date_${t}>`);

    }

}
