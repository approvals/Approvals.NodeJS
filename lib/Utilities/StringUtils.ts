export class StringUtils{

    static replaceAll(text: string, find: string, replace: string) {
        while (text.includes(find)){
            text = text.replace(find,replace);
        }
        return text;
    }
}
