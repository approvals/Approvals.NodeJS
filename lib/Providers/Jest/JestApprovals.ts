const StringWriter = require("../../StringWriter");
const approvals = require("../../Approvals");
import { Options } from "../../Core/Options";
import { getJestNamer } from "./JestNamer";


export function verify(sut: any, options?: Options): void {
    const config = approvals.getConfig();
    config.reporters = ["diffmerge"];
    options = options || new Options()
    const writer = new StringWriter(config,  `${sut}`, options.forFile().getFileExtension());
    approvals.verifyWithControl(getJestNamer(), writer, null, config);
}

export function verifyAsJson(data: any, options?: Options): void {
    const text = JSON.stringify(data, null, "  ");
    options = options || new Options()
    options = options.forFile().withFileExtention(".json")
    verify(text, options);
}
