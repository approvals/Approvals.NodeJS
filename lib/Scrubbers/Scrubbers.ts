import {DateScrubber} from "./DateScrubber";

export type Scrubber = (t: string) => string;

type ReplacementFunction = (index: number) => string;

function _replaceRegex(text: string, regex: RegExp, replacement: ReplacementFunction): string {
    const capturedGuids: string[] = []

    const result = text.replace(regex, match => {
        if (capturedGuids.indexOf(match) < 0) {
            capturedGuids.push(match);
        }

        const index = capturedGuids.indexOf(match) + 1;
        return replacement(index);
    });
    return result;
}

export class Scrubbers {

    static createReqexScrubber(regex: RegExp, replacement: (string | ReplacementFunction)): Scrubber {
        if (typeof replacement === 'function') {
            return t => _replaceRegex(t, regex, replacement as ReplacementFunction)
        } else {
            const replacementString = replacement as string;
            return t => _replaceRegex(t, regex, () => replacementString)
        }
    }

    static createGuidScrubber(): Scrubber {
        const guidReqex: RegExp = /[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/ig
        return this.createReqexScrubber(guidReqex, t => `<guid_${t}>`);
    }

    /**
     * This method exists as a convenient way to get an example scrubber for you to use.
     * To use this template, simply inline the method in your IDE.
     */
    static templates = class {
        static regexScrubberWithLambda(): Scrubber {
            return Scrubbers.createReqexScrubber(/your pattern here: [a-zA-Z]+d{4}/ig, t => `<your replacement_${t}>`);
        }

        static regexScrubber(): Scrubber {
            return Scrubbers.createReqexScrubber(/your pattern here: [a-zA-Z]+d{4}/ig, `<your replacement>`);
        }

        static dateScrubber(): Scrubber {
            return DateScrubber.getScrubberFor("2014/05/13 16:30:59.786");
        }
    }
}
