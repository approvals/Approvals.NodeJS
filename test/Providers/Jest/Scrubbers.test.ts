import {describe, expect, test} from "@jest/globals";
import {verify, verifyAsJson} from "../../../lib/Providers/Jest/JestApprovals";
import {Options} from "../../../lib/Core/Options";
import {StringUtils} from "../../../lib/Utilities/StringUtils";
import {Scrubbers} from "../../../lib/Scrubbers/Scrubbers";
import {DateScrubber} from "../../../lib/Scrubbers/DateScrubber";


describe("Scubbers", () => {
    test("scrub", () => {
        const text = "[1, 2, 12, 21, 121, 131, 222]";

        verify(text, new Options().withScrubber(t => StringUtils.replaceAll(t, "2", "two")));
    });
    test('guids', () => {
        const text = {
            people: [{id: "58f471f1-8b1f-413c-8971-21cb23bfc8f2", name: "Eve"},
                {
                    id: "24775957-3125-42db-ac6d-0abf142685b9",
                    name: "Kane",
                    mother: "58f471f1-8b1f-413c-8971-21cb23bfc8f2"
                }, {
                    id: "45365d5b-6cfc-430c-b842-36fbcd21470e",
                    name: "Kane",
                    mother: "58f471f1-8b1f-413c-8971-21cb23bfc8f2"
                }]
        };
        verifyAsJson(text, new Options().withScrubber(Scrubbers.createGuidScrubber()));
    });
    test('regex', () => {
        const text = "Here is some (｡◕ ∀ ◕｡) text";
        verifyAsJson(text, new Options().withScrubber(Scrubbers.createReqexScrubber(/\([\s\S]+\)/ig, "crazy")));
    });
    test('dates', () => {
        for (let format of DateScrubber.getSupportedFormats()) {
            for (let example of format.examples) {
                const scrubber = DateScrubber.getScrubberFor(example);
                expect(scrubber(example)).toBe("<date_1>")
            }
        }
    });

});
