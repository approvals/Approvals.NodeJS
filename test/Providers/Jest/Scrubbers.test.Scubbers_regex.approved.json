"Here is some crazy text"
